File: 2015/day-22/day-22-wizard-simulator-20XX.js
console.log("Part One:", partOne());
console.log("Part Two:", partTwo());

function partOne() {
  const game = new Game();
  game.playAllGames();
  return game.leastManaSpent;
}

function partTwo() {
  const game = new Game(true);
  game.playAllGames();
  return game.leastManaSpent;
}

class Game {
  constructor(isPartTwo = false) {
    this.wizard = new Wizard({ hitPoints: 50, mana: 500 });
    this.boss = new Boss({ hitPoints: 58, damageAmount: 9 });
    this.leastManaSpent = Infinity;
    this.isPartTwo = isPartTwo;
  }

  playAllGames(wizard = this.wizard, boss = this.boss) {
    for (let i = 0; i < wizard.spells.length; i++) {
      let spellActive = false;
      for (let j = 0; j < wizard.activeSpells.length; j++) {
        if (wizard.activeSpells[j].duration > 1 && i === wizard.activeSpells[j].idx) {
          spellActive = true;
        }
      }
      if (spellActive || wizard.spells[i].cost > wizard.mana) continue;

      const newWizard = wizard.duplicate();
      const newBoss = boss.duplicate();

      this.playRound(newWizard, newBoss, i);

      if (newBoss.hitPoints <= 0) {
        this.leastManaSpent = Math.min(this.leastManaSpent, newWizard.spent);
      }

      if (this.shouldContinue(newWizard, newBoss)) this.playAllGames(newWizard, newBoss);
    }
  }

  playRound(wizard, boss, spellIdx) {
    if (this.isPartTwo) wizard.hitPoints--;
    wizard.runEffects(boss);
    wizard.attack(boss, spellIdx);
    wizard.runEffects(boss);
    boss.attack(wizard);
  }

  shouldContinue(wizard, boss) {
    const wizardMinHp = this.isPartTwo ? 1 : 0;
    return wizard.hitPoints > wizardMinHp && boss.hitPoints > 0 && wizard.spent < this.leastManaSpent;
  }
}

class Player {
  constructor(hitPoints) {
    this.hitPoints = hitPoints;
    this.spent = 0;
    this.armor = 0;
    this.activeSpells = [];
  }

  takeDamage(amount) {
    this.hitPoints -= Math.max(1, amount - this.armor);
  }
}

class Spell {
  constructor({ name, cost, startFn, effectFn, endFn, duration }) {
    this.name = name;
    this.cost = cost;
    this.start = startFn;
    this.effect = effectFn;
    this.end = endFn;
    this.duration = duration;
  }
}

const spells = [
  new Spell({
    name: "Magic Missile",
    cost: 53,
    effectFn: (_wizard, boss) => boss.takeDamage(4),
  }),
  new Spell({
    name: "Drain",
    cost: 73,
    effectFn: (wizard, boss) => {
      boss.takeDamage(2);
      wizard.hitPoints += 2;
    },
  }),
  new Spell({
    name: "Shield",
    cost: 113,
    startFn: (wizard, _boss) => (wizard.armor += 7),
    effectFn: (_wizard, _boss) => { },
    endFn: (wizard, _boss) => (wizard.armor -= 7),
    duration: 6,
  }),
  new Spell({
    name: "Poison",
    cost: 173,
    effectFn: (_wizard, boss) => boss.takeDamage(3),
    duration: 6,
  }),
  new Spell({
    name: "Recharge",
    cost: 229,
    effectFn: (wizard, _boss) => (wizard.mana += 101),
    duration: 5,
  }),
];

class Wizard extends Player {
  constructor({ hitPoints, mana }) {
    super(hitPoints);
    this.spells = spells;
    this.mana = mana;
  }

  attack(boss, spellIdx) {
    const spell = this.spells[spellIdx];
    this.spent += spell.cost;
    this.mana -= spell.cost;

    if (spell.duration) {
      const newSpell = {
        idx: spellIdx,
        effect: spell.effect,
        duration: spell.duration,
      };
      if (spell.start) {
        spell.start(this, boss);
      }
      if (spell.end) {
        newSpell.end = spell.end;
      }
      this.activeSpells.push(newSpell);
    } else {
      spell.effect(this, boss);
    }
  }

  runEffects(boss) {
    for (const spell of this.activeSpells) {
      if (spell.duration > 0) {
        spell.effect(this, boss);
        spell.duration--;

        if (spell.end && spell.duration === 0) {
          spell.end(this, boss);
        }
      }
    }
  }

  duplicate() {
    const newWizard = new Wizard({ hitPoints: this.hitPoints, mana: this.mana });
    newWizard.spent = this.spent;
    newWizard.armor = this.armor;
    for (var i = 0; i < this.activeSpells.length; i++) {
      newWizard.activeSpells.push({ ...this.activeSpells[i] });
    }
    return newWizard;
  }
}

class Boss extends Player {
  constructor({ hitPoints, damageAmount }) {
    super(hitPoints);
    this.damageAmount = damageAmount;
  }

  attack(player, _spellIdx) {
    player.takeDamage(this.damageAmount);
  }

  duplicate() {
    var newBoss = new Boss({ hitPoints: this.hitPoints, damageAmount: this.damageAmount });
    newBoss.spent = this.spent;
    newBoss.armor = this.armor;
    return newBoss;
  }
}

File: 2015/day-22/day-22-wizard-simulator-20XX.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type Player struct {
	hitPoints int
	armor     int
	damage    int
	mana      int
}

func (p Player) attack(opponent *Player, amount int) {
	opponent.takeDamage(amount)
}

func (p *Player) takeDamage(amount int) {
	damageAmount := amount - p.armor
	if damageAmount < 1 {
		damageAmount = 1
	}
	p.hitPoints -= damageAmount
}

type Spell struct {
	name   string
	cost   int
	effect Effect
}

type Effect struct {
	counter int
	effect  func(EffectArgs)
}

type EffectArgs struct {
	player       *Player
	boss         *Player
	currentCount int
}

type Path struct {
	spells        []*Spell
	player        Player
	boss          Player
}

func (p *Path) Copy() Path {
	spells := []*Spell{}
	for _, s := range p.spells {
		spell := Spell{
			name: s.name,
			cost: s.cost,
			effect: s.effect,
		}
		spells = append(spells, &spell)
	}
	return Path{
		spells: spells,
		player: Player{ hitPoints: 50, mana: 500 },
		boss: getBoss(),
	}
}

func main() {
	p := Player{mana: 500, hitPoints: 50}
	boss := getBoss()
	spells := getSpells()

	pathQueue := []Path{
		{
			spells: []*Spell{},
			player: p,
			boss:   boss,
		},
	}
	minCost := math.MaxInt
	for len(pathQueue) > 0 {
		var latest Path
		latest, pathQueue = pathQueue[0], pathQueue[1:]
		if win, complete := playGame(&latest); complete && win {
			var sum int
			for _, s := range latest.spells {
				sum += s.cost
			}
			if sum < minCost {
				minCost = sum
				fmt.Println("sum:", sum)
			}

		} else {
			for _, s := range latest.spells {
				fmt.Printf("%v\n", s)
			}
			for _, s := range spells {
				p := latest.Copy()
				p.spells = append(p.spells, &s)
				pathQueue = append(pathQueue, p)
			}
		}
	}
}

func playGame(path *Path) (win bool, complete bool) {
	spells := make([]*Spell, len(path.spells))
	copy(spells, path.spells)
	for i, s := range spells {
		// this is essentially a turn - we'll need to go back and hit the different
		// active effects first and again before the boss goes
		for _, prevSpell := range spells[:i]  {
			if prevSpell.effect.counter > 0 {
				prevSpell.effect.effect(EffectArgs{player: &path.player, boss: &path.boss, currentCount: prevSpell.effect.counter})
				prevSpell.effect.counter--
			}
		}
		s.effect.effect(EffectArgs{player: &path.player, boss: &path.boss, currentCount: s.effect.counter})
		if path.player.hitPoints < 0 {
			return false, true
		}
		if path.boss.hitPoints < 0 {
			return true, true
		}

		for _, prevSpell := range spells[:i + 1]  {
			if prevSpell.effect.counter > 0 {
				prevSpell.effect.effect(EffectArgs{player: &path.player, boss: &path.boss, currentCount: prevSpell.effect.counter})
				prevSpell.effect.counter--
			}
		}

		path.boss.attack(&path.player, path.boss.damage)
		if path.player.hitPoints < 0 {
			return false, true
		}
		if path.boss.hitPoints < 0 {
			return true, true
		}
	}

	return false, false
}

func getBoss() Player {
	f, _ := os.ReadFile("./day-22-data.txt")
	lines := strings.Split(string(f), "\n")
	hitPointsLine, damageLine := lines[0], lines[1]

	_, hitPointsStr, _ := strings.Cut(hitPointsLine, ": ")
	_, damageStr, _ := strings.Cut(damageLine, ": ")

	hitPoints, _ := strconv.Atoi(hitPointsStr)
	damage, _ := strconv.Atoi(damageStr)

	return Player{
		hitPoints: hitPoints,
		damage:    damage,
	}
}

func getSpells() []Spell {
	return []Spell{
		{
			name: "Magic Missile",
			cost: 53,
			effect: Effect{
				counter: 0,
				effect: func(e EffectArgs) {
					e.player.mana -= 53
					e.player.attack(e.boss, 4)
				},
			},
		},
		{
			name: "Drain",
			cost: 73,
			effect: Effect{
				counter: 0,
				effect: func(e EffectArgs) {
					e.player.mana -= 73
					e.player.attack(e.boss, 2)
					e.player.hitPoints += 2
				},
			},
		},
		{
			name: "Shield",
			cost: 113,
			effect: Effect{
				counter: 6,
				effect: func(e EffectArgs) {
					if e.currentCount == 6 {
						e.player.mana -= 113
						e.player.armor += 7
					} else if e.currentCount == 0 {
						e.player.armor -= 7
						if e.player.armor < 0 {
							e.player.armor = 0
						}
					}
				},
			},
		},
		{
			name: "Poison",
			cost: 173,
			effect: Effect{
				counter: 6,
				effect: func(e EffectArgs) {
					e.player.attack(e.boss, 3)
				},
			},
		},
		{
			name: "Recharge",
			cost: 229,
			effect: Effect{
				counter: 5,
				effect: func(e EffectArgs) {
					e.player.mana += 101
				},
			},
		},
	}
}
